.github/scripts/transform_dbt_test_results.py

```diff
@@ -0,0 +1,1082 @@
+            "failing_rows": self.failing_rows,
+        }
+
+    @classmethod
+    def from_dict(cls, result_dict: typing.Dict) -> "TestResult":
+        """Deserialize a TestResult object from a dictionary."""
+        return TestResult(
+            name=result_dict["name"],
+            table_name=result_dict["table_name"],
+            status=Status(result_dict["status"]),
+            description=result_dict["description"],
+            elapsed_time=result_dict["elapsed_time"],
+            failing_rows=result_dict["failing_rows"],
+        )
+
+    def split_by_township(self) -> typing.List["TownshipTestResult"]:
+        """Split out this TestResult object into one or more TestResults
+        based on the township code of each failing row. If there are no failing
+        rows, or if all the failing rows"""
+        # Split out the failing rows by township so that we know which
+        # townships are represented in this test's failures
+        failing_rows_by_township = {}
+        for row in self.failing_rows:
+            township_code = row[TOWNSHIP_FIELD]
+            if not failing_rows_by_township.get(township_code):
+                failing_rows_by_township[township_code] = []
+            failing_rows_by_township[township_code].append(row)
+
+        # These kwargs are shared by all TownshipTestResults, regardless of
+        # township code or failure status
+        base_kwargs = {
+            "name": self.name,
+            "table_name": self.table_name,
+            "status": self.status,
+            "description": self.description,
+            "elapsed_time": self.elapsed_time,
+        }
+
+        # If we have any failing rows, split out separate TownshipTestResult
+        # objects for each township/row group; otherwise, just create one
+        # object with a null township mapped to the passing test
+        if failing_rows_by_township:
+            return [
+                TownshipTestResult(
+                    township_code=township_code,
+                    failing_rows=rows,
+                    **base_kwargs,
+                )
+                for township_code, rows in failing_rows_by_township.items()
+            ]
+        return [
+            TownshipTestResult(
+                township_code=None, failing_rows=[], **base_kwargs
+            )
+        ]
+
+
+class TownshipTestResult(TestResult):
+    """A variant of TestResult for a test whose results all share the same
+    township. Note that township_code is only present in the case of failing
+    tests; since passing tests have no township (or, thinking of it
+    differently, passing tests encompass all of the townships), the
+    township_code will always be None in the case of a passing test."""
+
+    def __init__(
+        self, township_code: typing.Optional[str], *args, **kwargs
+    ) -> None:
+        self.township_code = township_code
+        super().__init__(*args, **kwargs)
+
+
+class TestCategory:
```

Comment:
This class was adapted from the previous version of the `format_dbt_test_failures` script, where it was called [`FailedTestGroup`](https://github.com/ccao-data/data-architecture/blob/66c85bfa973d163bd29c4e41f08ecf58eafff08d/.github/scripts/format_dbt_test_failures.py#L95). The main difference between this version and the old version of the class is that, like `TestResult`, `TestCategory` can store results for passing tests in addition to failing tests.